### ISSUE TITLE
Handle byte slices and buffers more fruitfully

### ISSUE THREAD

**Issue by tomwhoiscontrary** [maxbrunsfeld/counterfeiter#30]
Fakes save the arguments to calls in a pretty straightforward way: value types are copied, and references to reference types are copied. This works fine for value types, and for reference types in the great majority of cases. However, it falls down a bit where the argument is a reference type which is subsequently mutated, and where you care about its value at the moment of the call, rather than its identity. For example, a byte slice being used as a buffer while working through a stream, or a bytes.Buffer (I struggle to think of any more examples!).

By way of example, this morning i was counterfeiting an interface with this method:

```go
PutBlock(container, name, blockID string, chunk []byte) error
```

I was testing code that 'put' several different 'blocks', one after the other. Let's say that i was uploading the alphabet in chunks of three letters. One test looked like:

```go
_, _, _, chunk := blobClient.PutBlockArgsForCall(0)
Expect(chunk).To(Equal([]byte{'a', 'b', 'c'}))
```

This failed; according to the fake, the chunk was {'y', 'z'}. That's because the fake saved the pointer to the slice, and the slice was subsequently filled with more data. It would have been really useful if the fake had somehow saved the {'a', 'b', 'c'} that was in the slice when it was passed.

I ended up modifying the fake by hand to save a copy of the chunk:

```go
copiedChunk := make([]byte, len(chunk))
copy(copiedChunk, chunk)
// ...
fake.putBlockArgsForCall = append(fake.putBlockArgsForCall, struct {
	container string
	name      string
	blockID   string
	chunk     []byte
}{container, name, blockID, copiedChunk})
```

Of course, this will be wiped away if i ever regenerate the fakes.

It would be great if there was a way to fix this properly. Capturing all byte slices as copies would be a breaking change, so perhaps a bad idea. But could there be some way to specify it?

**Comment by tjarratt** [comment]
This is a really interesting issue. I thought about this for a while, and I'm still not entirely sure where I stand on it. I'm glad you took the time to write this up and describe the issue so eloquently, @tomwhoiscontrary. Thank you.

At a certain point, adding more behavior to counterfeiter fakes (or really, test doubles in general) is dangerous. It's dangerous because this is untested behavior that you rely on, and your test doubles can behave in ways that are wildly different than your actual objects. I'd be a little concerned about adding this to counterfeiter fakes because it's increasing the surface area of what they do and it's hard to reason about things that do too much.

However, it's also a little surprising that slices are captured as a pointer. Is it only byte slices? My intuition tells me that *any* slice would be captured as a pointer. If that's true, then I'd be supportive of a PR that makes it so slices are always copied first.

My gut tells me that capturing the args is *intended* to capture the values as they were when the function was invoked. I'm having a hard time believing that someone would want to rely on the current behavior, because it seems so inconsistent and wrong the more I think about it.

**Comment by tomwhoiscontrary** [comment]
I assume any slice would be captured the same way.

I had thought that switching to copying slices would be a breaking change, but now i don't think it is, really. It would break tests which compare the captured arguments by identity, but AFAIK there's no direct way to do that: you can't compare slices for equality as you can real pointers, and you can't extract a pointer to the underlying array from a slice. So, a user would have to concoct an indirect test for identity, say by assigning a value to one slice, and seeing another slice change. That really feels like something we don't need to worry about!

Rather, i imagine users are taking the returned slice and looking at its contents somehow, which would continue to work fine.

I'll see if i can come up with a PR which copies slices.

I note that this won't do anything for bytes.Buffers.

**Comment by tjarratt** [comment]
Fixed by merging #33.
